File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/map/info/NOverlayQuery.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.map.info

import dev.note11.flutter_naver_map.flutter_naver_map.model.enum.NOverlayType

internal data class NOverlayQuery(val info: NOverlayInfo, val methodName: String) {
    val query: String get() = encode()

    // ----- Encode -----
    private fun encode(): String =
        listOf(info.type.toString(), info.id, methodName).joinToString(SEPARATOR)

    companion object {
        fun fromQuery(query: String): NOverlayQuery = decode(query)

        // ----- Decode -----
        private fun decode(string: String): NOverlayQuery {
            val split = string.split(SEPARATOR)
            val type = split.first()
            val method = split.last()
            val id = split.drop(1).dropLast(1).joinToString(SEPARATOR)

            return NOverlayQuery(
                NOverlayInfo(NOverlayType.fromString(type), id),
                methodName = method
            )
        }

        private const val SEPARATOR = "\""
    }
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/model/map/info/NOverlayQuery.swift
internal struct NOverlayQuery {
    let info: NOverlayInfo
    let methodName: String

    var query: String {
        encode()
    }

    static func fromQuery(_ query: String) -> NOverlayQuery {
        decode(string: query)
    }

    // ----- Encode / Decode -----

    private func encode() -> String {
        [info.type.rawValue, info.id, methodName].joined(separator: NOverlayQuery.SEPARATOR)
    }

    private static func decode(string: String) -> NOverlayQuery {
        let split = string.components(separatedBy: NOverlayQuery.SEPARATOR)
        let type = split.first!
        let method = split.last!
        let id = split.dropFirst().dropLast().joined(separator: NOverlayQuery.SEPARATOR)
        return NOverlayQuery(
                info: NOverlayInfo(type: NOverlayType(rawValue: type)!, id: id),
                methodName: method
        )
    }

    private static let SEPARATOR = "\""
}